File: greeting/python/src/greeting/greeting.py
from __future__ import annotations


def greet(name: str | list[str | None] | None) -> str:
    if isinstance(name, list):
        names = [n if n is not None else "my friend" for n in name]
    else:
        names = [name if name is not None else "my friend"]

    normals: list[str] = []
    shouts: list[str] = []
    for n in names:
        if _is_shout(n):
            shouts.append(n)
        else:
            normals.append(n)

    if not shouts:
        return _normal_greeting(normals)
    if not normals:
        return _shout_greeting(shouts)
    return f"{_normal_greeting(normals)}. AND {_shout_greeting(shouts)}"


def _normal_greeting(names: list[str]) -> str:
    if len(names) == 1:
        return f"Hello, {names[0]}."
    if len(names) == 2:
        return f"Hello, {names[0]} and {names[1]}"
    head = ", ".join(names[:-1])
    return f"Hello, {head}, and {names[-1]}"


def _shout_greeting(names: list[str]) -> str:
    return f"HELLO {' '.join(names)}!"


def _is_shout(name: str) -> bool:
    return any(c.isalpha() for c in name) and name == name.upper()

File: greeting/csharp/src/Greeting/Greeting.cs
namespace Greeting;

public static class Greeter
{
    public static string Greet(string? name)
    {
        return GreetMany(new[] { name ?? "my friend" });
    }

    public static string Greet(string?[] names)
    {
        var resolved = new string[names.Length];
        for (var i = 0; i < names.Length; i++)
        {
            resolved[i] = names[i] ?? "my friend";
        }
        return GreetMany(resolved);
    }

    private static string GreetMany(string[] names)
    {
        var normals = new List<string>();
        var shouts = new List<string>();
        foreach (var name in names)
        {
            if (IsShout(name)) shouts.Add(name);
            else normals.Add(name);
        }

        if (shouts.Count == 0) return NormalGreeting(normals);
        if (normals.Count == 0) return ShoutGreeting(shouts);
        return $"{NormalGreeting(normals)}. AND {ShoutGreeting(shouts)}";
    }

    private static string NormalGreeting(List<string> names)
    {
        if (names.Count == 1) return $"Hello, {names[0]}.";
        if (names.Count == 2) return $"Hello, {names[0]} and {names[1]}";
        var head = string.Join(", ", names.GetRange(0, names.Count - 1));
        return $"Hello, {head}, and {names[^1]}";
    }

    private static string ShoutGreeting(List<string> names)
    {
        return $"HELLO {string.Join(" ", names)}!";
    }

    private static bool IsShout(string name)
    {
        var hasLetter = false;
        foreach (var c in name)
        {
            if (char.IsLetter(c))
            {
                hasLetter = true;
                if (char.IsLower(c)) return false;
            }
        }
        return hasLetter;
    }
}
